text/0045-docker-proxy.md

```diff
@@ -0,0 +1,167 @@
+- Feature Name: docker_proxy
+- Start Date: 2019-07-01
+- RFC PR: https://github.com/hyperledger/sawtooth-rfcs/pull/45
+- Sawtooth Issue: N/A
+
+# Summary
+[summary]: #summary
+
+Sawtooth has containerized build and test processes for core and all
+components. Running Docker containers behind a network proxy requires
+special handling so that the containers can find external routes to download
+packages and build dependencies, for example. There are a number of
+ways to configure these environment for proxy support. It is desirable that we
+adopt a single consistent approach across the large body of Docker and
+Docker Compose files in the Sawtooth project.
+
+This RFC defines the canonical approach for supporting proxies for
+Sawtooth containers using Docker 18.x and Docker Compose 1.24.x with
+Ubuntu 16.04 and 18.04 and serves as guidelines to resolve these issues.
+
+# Motivation
+[motivation]: #motivation
+
+Sawtooth repositories make extensive use of Docker containers to build and
+test the software as well as provide turn key examples. Those containers
+typically need to fetch external packages or make other outbound network
+calls. The containers need special configuration to be executed behind
+network proxies. There are multiple ways to provide that proxy configuration
+and it is desirable that the Sawtooth repositories standardize on a single
+approach. The challenge is further complicated by different abilities and
+constraints of different versions of Docker and Ubuntu.
+
+Starting with Ubuntu 18 (Bionic), `gpg` , invoked through `apt-key adv` ,
+now ignores `http_proxy` and `https_proxy` environment variables, and ignores
+the command line options
+`apt-key adv --keyserver-options http-proxy=<proxy server url>`
+and `https-proxy=<proxy server url>` . The only method remaining is with
+cumbersome apt command-specific configuration file settings in
+`/etc/apt/apt.conf.d/proxy` .
+
+
+# Goals
+[goals]: #goals
+
+* Have Docker and Docker Compose files and shell scripts that work in explicit
+  proxy, transparent proxy, and non-proxy environments
+* Have a uniform method to access proxy servers across Hyperledger Sawtooth
+  projects * Minimize Docker and script clutter with unnecessary tests and
+  settings
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
```

Comment:
I'd prefer that this link to the official docker documentation rather than try to duplicate it in case the process changes in the future.

Reply:
I added a link to the official Docker document. It does not cover the special case of key servers though.

Reply:
Can you move the docker documentation link to this section instead? It's relevant in this section, but not to the one below.

Reply:
Yes.